cdap-data-fabric/src/test/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTest.java

```diff
@@ -0,0 +1,330 @@
+  public void testDecodeIncomplete() {
+    // key can be in any order... partitioning dictates the order of fields in row key
+    PartitionKey key = PartitionKey.builder()
+      .addIntField("i", 42)
+      .addLongField("l", 17L)
+      .addStringField("s", "x")
+      .build();
+    byte[] rowKey = PartitionedFileSetDataset.generateRowKey(key, PARTITIONING_1);
+    PartitionedFileSetDataset.parseRowKey(rowKey, PARTITIONING_2);
+  }
+
+  @Test
+  @Category(SlowTests.class)
+  public void testAddRemoveGetPartitions() throws Exception {
+
+    final PartitionedFileSet dataset = getInstance("pfs");
+
+    final PartitionKey[][][] keys = new PartitionKey[4][4][4];
+    final String[][][] paths = new String[4][4][4];
+    final Map<PartitionKey, String> allPartitions = Maps.newHashMap();
+
+    // add a bunch of partitions
+    for (int s = 0; s < 4; s++) {
+      for (int i = 0; i < 4; i++) {
+        for (int l = 0; l < 4; l++) {
+          final PartitionKey key = PartitionKey.builder()
+            .addField("s", String.format("%c-%d", 'a' + s, s))
+            .addField("i", i * 100)
+            .addField("l", 15L - 10 * l)
+            .build();
+          final String path = String.format("%s-%d-%d", s, i, l);
+          newTransactionExecutor((TransactionAware) dataset).execute(new TransactionExecutor.Subroutine() {
+            @Override
+            public void apply() throws Exception {
+              dataset.addPartition(key, path);
+            }
+          });
+          keys[s][i][l] = key;
+          paths[s][i][l] = path;
+          allPartitions.put(key, path);
+        }
+      }
+    }
+
+    // validate getPartition with exact partition key
+    for (int s = 0; s < 4; s++) {
+      for (int i = 0; i < 4; i++) {
+        for (int l = 0; l < 4; l++) {
+          final PartitionKey key = keys[s][i][l];
+          final String path = paths[s][i][l];
+          newTransactionExecutor((TransactionAware) dataset).execute(new TransactionExecutor.Subroutine() {
+            @Override
+            public void apply() throws Exception {
+              Assert.assertEquals(path, dataset.getPartition(key));
+            }
+          });
+          // also test getPartitionPaths() and getPartitions() for the filter matching this
+          @SuppressWarnings({"unchecked", "unused"})
+          boolean success = testFilter(dataset, allPartitions,
+                                       PartitionFilter.builder()
+                                         .addValueCondition("l", key.getField("l"))
+                                         .addValueCondition("s", key.getField("s"))
+                                         .addValueCondition("i", key.getField("i"))
+                                         .build());
+        }
+      }
+    }
+
+    // test whether query works without filter
+    testFilter(dataset, allPartitions, null);
+
+    // generate an list of partition filters with exhaustive coverage
```

Comment:
this generates about 2700 different filters. And we test all of them five times. The whole test takes about 8 seconds on my laptop, therefore I marked it as a slow test. It is still useful to test all these combinations, even though I am sure there is some redundancy between them, it helped me find a handful of bugs for border conditions.